apps/src/blockly/addons/cdoFieldButton.js

```diff
@@ -45,25 +45,34 @@ export default class CdoFieldButton extends GoogleBlockly.Field {
   }
 
   /**
+   * Used to update the value of a field.
    * @override
+   * @param newValue The value to be saved.
    */
-  getValue() {
-    return String(this.value_);
+  doValueUpdate_(newValue) {
+    if (this.value_ !== newValue) {
+      if (this.changeHandler_) {
+        const override = this.changeHandler_(newValue);
+        if (override !== undefined) {
+          newValue = override;
+        }
+      }
+      this.value_ = newValue;
+    }
+    this.isDirty_ = true; // Block needs to be re-rendered.
   }
 
   /**
+   * Validate the changes to the field's value before they are set.
    * @override
+   * @param newValue The value that is being validated
+   * @returns `String` or `null`
    */
-  setValue(value) {
-    if (this.value_ !== value) {
-      if (this.changeHandler_) {
-        const override = this.changeHandler_(value);
-        if (override !== undefined) {
-          value = override;
-        }
-      }
-      this.value_ = value;
+  doClassValidation_(newValue) {
+    if (newValue === null || newValue === undefined) {
+      return null;
     }
+    return String(newValue);
```

Comment:
Help me understand generics/typescript...looking at [the existing](https://github.com/google/blockly/blob/139bc299a061c3d9fc81e8026d929cbe5589bb55/core/field.ts#L1092) `doClassValidation_`, how does this differ?

Reply:
I'm wanted to ensure the value is stored as a String based on our conversation from a [past PR](https://github.com/code-dot-org/code-dot-org/pull/50834#discussion_r1171943575).  I saw that `T` [is defined with `any`](https://github.com/google/blockly/blob/139bc299a061c3d9fc81e8026d929cbe5589bb55/core/field.ts#L68), so I overrode this function. But I am really new to TypeScript so lmk if this sounds right!

Btw, I see that `FieldButton` is used once in [`blocks.js`](https://github.com/code-dot-org/code-dot-org/blob/e3733d4cf6baa8586981ca8b0a56569756e7e689/apps/src/p5lab/spritelab/blocks.js#L123), and `setValue` is always called with a string.

Reply:
Just to clarify - types, and all of TypeScript really only exists compile-time, by design. At runtime, it's just running compiled JS code. So the `as T` cast just tells TypeScript to treat this value as type `T` when referencing the return value of this function (in other TypeScript code). At runtime, this is effectively `return newValue;`. So wrapping this in `String()` does ensure that we're actually returning a String-ified version of the value here.

It's probably also worth mentioning that the Blockly codebase seems to be actively in the middle of a TypeScript migration right now (ex. types like 'AnyDuringMigration'). FWIW we had to recently rollback to a slightly older version of blockly due to a scrolling bug, and so the version we're on actually doesn't have generics in `Field`; I believe the type of the `value` in a Field is just `any`.